### PR TITLE
[Ready for review] 1.5.0 CSV Output

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,6 +4,10 @@ These classes provide a simple way to bootstrap a Slim application with authenti
 
 It is an abstraction of the [Slim Framework](http://slimframework.com/) and handles some stuff like output generation in different formats and authentication / acl handling.
 
+# Work in progress feature-branch!
+
+This branch is to develop the feature of a CSV response.
```
Comment (ghost) [inline README.md]: please remove this and add a proper feature description before merging this into master

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        foreach($identifiers as $key =>$value){
+            $newIdentifiers['identifier_' . $key] = $value;
+        }
+        $newIdentifiers = $this->_flatten($newIdentifiers);
+
+        $tmp = \array_merge($newIdentifiers, $data->getData());
+        $result[] = $tmp;
+    }
+
+    private function _flatten($origin, $namespace = null) {
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: missing doc block

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    /**
+     * @param   array   $data       Structured & flattened payload as 2D-array
+     *                                  array($payload1, $payload2, ...);
+     * @param   bool    $encloseAll Force enclosing every field (false)
+     *
+     * @return  bool|string         Returns the CSV or false in case of an error.
+     */
+    protected function _csvEncode($data, $encloseAll = false)
+    {
+        //TODO: implement this
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: remove todo

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+        foreach ($data as $first) {
+            if (false === \is_array($first)) {
+                continue;
+            }
+            $returnCsv[] = '# ' . \implode($this->_delimiter, \array_keys($first));
+            break;
+        }
+
+        foreach ($data as $entry) {
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php:204]: why do you iterate the $data two time? Can it be done in one loop?

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        $delimiter_esc = \preg_quote($this->_delimiter, '/');
+        $enclosure_esc = \preg_quote($this->_enclosure, '/');
+
+        $output = array();
+        foreach ($fields as $field) {
+            if ($field === null) {
+                $output[] = $this->_null;
+                continue;
+            } else if (true === \is_array($field)) {
+                // TODO: Add warning: "Payload not flattened!"
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: todo?!

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        $result = array();
+
+        if (true === \is_array($data)) {
+            $this->_normalizeAll($data);
+            $identifiers = null;
+            foreach ($data as $entry) {
+                $this->_buildStructure($entry, $entry->getIdentifiers(), 0, $result);
+            }
+        } else {
+            $this->_buildStructure($this->_normalizeOne($data), $data->getIdentifiers(), 0, $result);
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: please put the _normalizeOne call into an extra line

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+        $this->_response->setStatus($statusCode);
+        $this->_response->setBody($body);
+    }
+
+    /**
+     * @param SlimBootstrap\DataObject $data  DataObject
+     * @param array|null $keys
+     * @return array
+     */
+    private function _normalizeOne(SlimBootstrap\DataObject &$data, $keys = null){
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: please avoid using parameter reverences

see other functions too

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        if(null === $keys){
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: enter space between if and (
check code below too

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            $keys[] = \array_keys($data->getData());
+        }
+        $keys = \array_fill_keys($keys, null);
+
+        if (\count(\array_intersect_key($keys, $data->getData())) !== \count($keys)){
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: please add a small comment, which explains why you need to update original data.

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+        } else {
+            return;
+        }
+    }
+
+    /**
+     * @param SlimBootstrap\DataObject[] $data
+     * @throws SlimBootstrap\Exception
+     */
+    private function _normalizeAll(&$data){
```
Comment (pahrens) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: please avoid using parameter reverences

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+    /**
+     * @param   array $fields Structured 2+-dimensional-data array
+     * @param   bool $encloseAll Force enclosing every field (false)
+     * @return bool|string Returns the line for $fields or false in case of an error.
+     *
+     * @throws SlimBootstrap\Exception
+     * Adapted from:
+     * @see http://php.net/manual/en/function.fputcsv.php#87120
+     */
+    private function _dataSetToLine(array $fields, $encloseAll = false) {
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: maybe better function would be: _buildCsvLine

**Comment by ghost** [comment]
unit test look broken

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+        if(null === $keys){
+            $keys = \array_keys($data->getData());
+        }
+
+        $keys = \array_fill_keys($keys, null);
+
+        if (\count(
+                \array_intersect_key(
+                    $keys,
+                    $data->getData()
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: please move count call out of the if condition, so the condition is better readable

File: src/SlimBootstrap/ResponseOutputWriter/Csv.php [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+     * @return SlimBootstrap\DataObject[]   Normalized DataObjects
+     *
+     * @throws SlimBootstrap\CSVEncodingException
+     */
+    protected function _normalizeAll($data)
+    {
+        $keys           = array();
+        $identifierKeys    = null;
+
+        foreach ($data as &$entry) {
```
Comment (ghost) [inline src/SlimBootstrap/ResponseOutputWriter/Csv.php]: please do not use variable references

**Pull request merged**